Ability to specify directory that libasyncProfiler is exported to, to support /tmp partitions with noexec set

## Is your feature request related to a problem?
Following CIS benchmarks, in our environments, /tmp is mounted as noexec for safety. `java.io.tmpdir` is left at the default `/tmp` as our applications do not need to execute anything that is stored in /tmp (e.g. user's file uploads). With this configuration, turning on `elastic.apm.profiling_inferred_spans_enabled` results in the following stacktrace:
```
2020-06-09 19:23:18.382 [elastic-apm-sampling-profiler] ERROR co.elastic.apm.agent.profiler.SamplingProfiler - Stopping profiler
java.lang.UnsatisfiedLinkError: /tmp/libasyncProfiler-linux-x64-99e9233a3c10a1a528595299ac634422.so: /tmp/libasyncProfiler-linux-x64-99e9233a3c10a1a528595299ac634422.so: failed to map segment from shared object: Operation not permitted
        at java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)
        at java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2430)
```
After looking at [this](https://github.com/elastic/apm-agent-java/blob/master/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfiler.java#L84) and [this](https://github.com/elastic/apm-agent-java/blob/master/apm-agent-core/src/main/java/co/elastic/apm/agent/util/IOUtils.java#L223), testing a workaround of setting `-Djava.io.tmpdir` to a partition that has exec permissions did solve the issue, but that also applies to our applications, which we would not want to have such permissions.

## Describe the solution you'd like
A separate configuration property for setting the path for libasyncProfiler.so to be exported to, to support a partition with exec partitions without affecting the configuration of the profiled application, e.g. `-Delastic.apm.profiling_inferred_spans_lib_directory=/var/tmp`

## Describe alternatives you've considered
Setting `java.io.tmpdir` as described above

## Additional context
N/A

> Sounds like a good idea! Do you want to take a stab at the implementation?

Took a while to get round to it, sorry, but I've put out a PR: #1350

> We should also show a friendlier error message if `java.io.File#canExecute` returns false, suggesting to set the new property.

I did some testing with `java.io.File#canExecute` but it always returned false, as the libasyncProfiler.so file itself is not executable (and doesn't need to be). If you think that should be changed for this purpose, I can try that. As it stands, the PR catches the UnsatisfiedLinkError and re-throws it, nested in an IllegalStateException that gives some info on the new property.

It's my first PR in this repo so no doubt it's not quite right - happy to make any changes obviously

This is what the output looks like by the way: 

```
2020-08-19 20:59:49,434 [main] INFO  co.elastic.apm.agent.impl.ElasticApmTracer - Tracer switched to RUNNING state
2020-08-19 20:59:49,475 [elastic-apm-sampling-profiler] ERROR co.elastic.apm.agent.profiler.SamplingProfiler - Stopping profiler
java.lang.IllegalStateException: It is likely that /tmp is not an executable location. Consider setting the profiling_inferred_spans_lib_directory property to a directory on a partition that allows execution
	at co.elastic.apm.agent.profiler.asyncprofiler.AsyncProfiler.getInstance(AsyncProfiler.java:78) ~[?:?]
	at co.elastic.apm.agent.profiler.SamplingProfiler.profile(SamplingProfiler.java:336) ~[?:?]
	at co.elastic.apm.agent.profiler.SamplingProfiler.run(SamplingProfiler.java:317) [?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.UnsatisfiedLinkError: /tmp/libasyncProfiler-linux-x64-12c3701c01997ba08ef2c3f5f4bbd39a.so: /tmp/libasyncProfiler-linux-x64-12c3701c01997ba08ef2c3f5f4bbd39a.so: failed to map segment from shared object: Operation not permitted
	at java.lang.ClassLoader$NativeLibrary.load0(Native Method) ~[?:?]
	at java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2430) ~[?:?]
	at java.lang.ClassLoader$NativeLibrary.loadLibrary(ClassLoader.java:2487) ~[?:?]
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:2684) ~[?:?]
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:2617) ~[?:?]
	at java.lang.Runtime.load0(Runtime.java:767) ~[?:?]
	at java.lang.System.load(System.java:1834) ~[?:?]
	at co.elastic.apm.agent.profiler.asyncprofiler.AsyncProfiler.loadNativeLibrary(AsyncProfiler.java:93) ~[?:?]
	at co.elastic.apm.agent.profiler.asyncprofiler.AsyncProfiler.getInstance(AsyncProfiler.java:76) ~[?:?]
	... 8 more
```

--- a/CHANGELOG.asciidoc
+++ b/CHANGELOG.asciidoc
@@ -50,7 +50,7 @@ endif::[]
   an empty string - {pull}1295[#1295]
 * Add <<metrics-micrometer, micrometer support>> - {pull}1303[#1303]
 * Add `profiling_inferred_spans_lib_directory` option to override the default temp directory used for exporting the async-profiler library.
-  This is useful for server-hardened environments where /tmp is often configured with noexec, leading to java.lang.UnsatisfiedLinkError errors - {pull}1350[#1350]
+  This is useful for server-hardened environments where `/tmp` is often configured with `noexec`, leading to `java.lang.UnsatisfiedLinkError` errors - {pull}1350[#1350]
 
 [float]
 ===== Bug fixes

--- a/CHANGELOG.asciidoc
+++ b/CHANGELOG.asciidoc
@@ -49,6 +49,8 @@ endif::[]
 * Enabling agent to work without attempting any communication with APM server, by allowing setting `server_urls` with
   an empty string - {pull}1295[#1295]
 * Add <<metrics-micrometer, micrometer support>> - {pull}1303[#1303]
+* Add `profiling_inferred_spans_lib_directory` option to override the default temp directory used for exporting the async-profiler library.
+  This is useful for server-hardened environments where /tmp is often configured with noexec, leading to java.lang.UnsatisfiedLinkError errors - {pull}1350[#1350]
 
 [float]
 ===== Bug fixes

--- a/apm-agent-core/src/main/java/co/elastic/apm/agent/util/IOUtils.java
+++ b/apm-agent-core/src/main/java/co/elastic/apm/agent/util/IOUtils.java
@@ -214,13 +214,14 @@ public class IOUtils {
      * Why it's synchronized : if the same JVM try to lock file, we got an java.nio.channels.OverlappingFileLockException.
      * So we need to block until the file is totally written.
      */
-    public static synchronized File exportResourceToTemp(String resource, String tempFileNamePrefix, String tempFileNameExtension) {
+    public static synchronized File exportResourceToDirectory(String resource, String parentDirectory, String tempFileNamePrefix,
+                                                              String tempFileNameExtension) {
         try (InputStream resourceStream = IOUtils.class.getResourceAsStream("/" + resource)) {
             if (resourceStream == null) {
                 throw new IllegalStateException(resource + " not found");
             }
             String hash = md5Hash(IOUtils.class.getResourceAsStream("/" + resource));
-            File tempFile = new File(System.getProperty("java.io.tmpdir"), tempFileNamePrefix + "-" + hash + tempFileNameExtension);
+            File tempFile = new File(parentDirectory, tempFileNamePrefix + "-" + hash + tempFileNameExtension);
             if (!tempFile.exists()) {
                 try (FileOutputStream out = new FileOutputStream(tempFile)) {
                     FileChannel channel = out.getChannel();

--- a/apm-agent-core/src/test/java/co/elastic/apm/agent/util/IOUtilsTest.java
+++ b/apm-agent-core/src/test/java/co/elastic/apm/agent/util/IOUtilsTest.java
@@ -56,6 +56,8 @@ import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 class IOUtilsTest  {
 
+    private final String tempDirectory = System.getProperty("java.io.tmpdir");
+
     @Test
     void readUtf8Stream() throws IOException {
         final CharBuffer charBuffer = CharBuffer.allocate(8);
@@ -159,8 +161,8 @@ class IOUtilsTest  {
     }
 
     @Test
-    void exportResourceToTemp() throws UnsupportedEncodingException, URISyntaxException {
-        File tmp = IOUtils.exportResourceToTemp("test.elasticapm.properties", UUID.randomUUID().toString(), "tmp");
+    void exportResourceToDirectory() throws UnsupportedEncodingException, URISyntaxException {
+        File tmp = IOUtils.exportResourceToDirectory("test.elasticapm.properties", tempDirectory, UUID.randomUUID().toString(), "tmp");
         tmp.deleteOnExit();
 
         Path referenceFile = Paths.get(IOUtilsTest.class.getResource("/test.elasticapm.properties").toURI());
@@ -170,24 +172,24 @@ class IOUtilsTest  {
     }
 
     @Test
-    void exportResourceToTempIdempotence() throws InterruptedException
+    void exportResourceToDirectoryIdempotence() throws InterruptedException
     {
         String destination = UUID.randomUUID().toString();
-        File tmp = IOUtils.exportResourceToTemp("test.elasticapm.properties", destination, "tmp");
+        File tmp = IOUtils.exportResourceToDirectory("test.elasticapm.properties", tempDirectory, destination, "tmp");
         tmp.deleteOnExit();
         long actual = tmp.lastModified();
         Thread.sleep(1000);
-        File after = IOUtils.exportResourceToTemp("test.elasticapm.properties", destination, "tmp");
+        File after = IOUtils.exportResourceToDirectory("test.elasticapm.properties", tempDirectory, destination, "tmp");
         assertThat(actual).isEqualTo(after.lastModified());
     }
 
     @Test
-    void exportResourceToTemp_throwExceptionIfNotFound() {
-        assertThatThrownBy(() -> IOUtils.exportResourceToTemp("nonexist", UUID.randomUUID().toString(), "tmp")).hasMessage("nonexist not found");
+    void exportResourceToDirectory_throwExceptionIfNotFound() {
+        assertThatThrownBy(() -> IOUtils.exportResourceToDirectory("nonexist", tempDirectory, UUID.randomUUID().toString(), "tmp")).hasMessage("nonexist not found");
     }
 
     @Test
-    void exportResourceToTempInMultipleThreads() throws InterruptedException, ExecutionException, IOException
+    void exportResourceToDirectoryInMultipleThreads() throws InterruptedException, ExecutionException, IOException
     {
         final int nbThreads = 10;
         final ExecutorService executorService = Executors.newFixedThreadPool(nbThreads);
@@ -199,7 +201,7 @@ class IOUtilsTest  {
             futureList.add(executorService.submit(() -> {
                 countDownLatch.countDown();
                 countDownLatch.await();
-                File file = IOUtils.exportResourceToTemp("test.elasticapm.properties", tempFileNamePrefix, "tmp");
+                File file = IOUtils.exportResourceToDirectory("test.elasticapm.properties", tempDirectory, tempFileNamePrefix, "tmp");
                 file.deleteOnExit();
                 return file;
             }));

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/ProfilingConfiguration.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/ProfilingConfiguration.java
@@ -170,7 +170,7 @@ public class ProfilingConfiguration extends ConfigurationOptionProvider {
             "The partition backing this location must be executable, however in some server-hardened environments, " +
             "noexec may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.\n" +
             "Set this property to an alternative directory (e.g. `/var/tmp`) to resolve this.\n" +
-            "If unset, the value of the `java.io.tmpdir` System property will be used.")
+            "If unset, the value of the `java.io.tmpdir` system property will be used.")
         .configurationCategory(PROFILING_CATEGORY)
         .dynamic(false)
         .tags("added[1.18.0]")

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/ProfilingConfiguration.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/ProfilingConfiguration.java
@@ -168,9 +168,9 @@ public class ProfilingConfiguration extends ConfigurationOptionProvider {
         .description("Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library " +
             "is exported to a temporary location and loaded by the JVM.\n" +
             "The partition backing this location must be executable, however in some server-hardened environments, " +
-            "noexec may be set on the standard /tmp partition, leading to java.lang.UnsatisfiedLinkError errors.\n" +
-            "Set this property to an alternative directory (e.g. '/var/tmp') to resolve this.\n" +
-            "If unset, the value of the \"java.io.tmpdir\" System property will be used.")
+            "noexec may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.\n" +
+            "Set this property to an alternative directory (e.g. `/var/tmp`) to resolve this.\n" +
+            "If unset, the value of the `java.io.tmpdir` System property will be used.")
         .configurationCategory(PROFILING_CATEGORY)
         .dynamic(false)
         .tags("added[1.18.0]")

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/ProfilingConfiguration.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/ProfilingConfiguration.java
@@ -34,7 +34,6 @@ import org.stagemonitor.configuration.ConfigurationOptionProvider;
 
 import java.util.Arrays;
 import java.util.List;
-import java.util.Optional;
 
 import static co.elastic.apm.agent.configuration.validation.RangeValidator.isInRange;
 import static co.elastic.apm.agent.configuration.validation.RangeValidator.min;
@@ -164,7 +163,7 @@ public class ProfilingConfiguration extends ConfigurationOptionProvider {
         .tags("added[1.15.0]", "internal")
         .buildWithDefault(TimeDuration.of("5s"));
 
-    private final ConfigurationOption<Optional<String>> profilerLibDirectory = ConfigurationOption.<String>stringOption()
+    private final ConfigurationOption<String> profilerLibDirectory = ConfigurationOption.<String>stringOption()
         .key("profiling_inferred_spans_lib_directory")
         .description("Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library " +
             "is exported to a temporary location and loaded by the JVM.\n" +
@@ -175,7 +174,7 @@ public class ProfilingConfiguration extends ConfigurationOptionProvider {
         .configurationCategory(PROFILING_CATEGORY)
         .dynamic(false)
         .tags("added[1.18.0]")
-        .buildOptional();
+        .build();
 
     public boolean isProfilingEnabled() {
         return profilingEnabled.get();
@@ -222,6 +221,6 @@ public class ProfilingConfiguration extends ConfigurationOptionProvider {
     }
 
     public String getProfilerLibDirectory() {
-        return profilerLibDirectory.get().orElse(System.getProperty("java.io.tmpdir"));
+        return profilerLibDirectory.isDefault() ? System.getProperty("java.io.tmpdir") : profilerLibDirectory.get();
     }
 }

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/ProfilingConfiguration.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/ProfilingConfiguration.java
@@ -168,7 +168,7 @@ public class ProfilingConfiguration extends ConfigurationOptionProvider {
         .description("Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library " +
             "is exported to a temporary location and loaded by the JVM.\n" +
             "The partition backing this location must be executable, however in some server-hardened environments, " +
-            "noexec may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.\n" +
+            "`noexec` may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.\n" +
             "Set this property to an alternative directory (e.g. `/var/tmp`) to resolve this.\n" +
             "If unset, the value of the `java.io.tmpdir` system property will be used.")
         .configurationCategory(PROFILING_CATEGORY)

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/ProfilingConfiguration.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/ProfilingConfiguration.java
@@ -34,6 +34,7 @@ import org.stagemonitor.configuration.ConfigurationOptionProvider;
 
 import java.util.Arrays;
 import java.util.List;
+import java.util.Optional;
 
 import static co.elastic.apm.agent.configuration.validation.RangeValidator.isInRange;
 import static co.elastic.apm.agent.configuration.validation.RangeValidator.min;
@@ -163,6 +164,19 @@ public class ProfilingConfiguration extends ConfigurationOptionProvider {
         .tags("added[1.15.0]", "internal")
         .buildWithDefault(TimeDuration.of("5s"));
 
+    private final ConfigurationOption<Optional<String>> profilerLibDirectory = ConfigurationOption.<String>stringOption()
+        .key("profiling_inferred_spans_lib_directory")
+        .description("Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library " +
+            "is exported to a temporary location and loaded by the JVM.\n" +
+            "The partition backing this location must be executable, however in some server-hardened environments, " +
+            "noexec may be set on the standard /tmp partition, leading to java.lang.UnsatisfiedLinkError errors.\n" +
+            "Set this property to an alternative directory (e.g. '/var/tmp') to resolve this.\n" +
+            "If unset, the value of the \"java.io.tmpdir\" System property will be used.")
+        .configurationCategory(PROFILING_CATEGORY)
+        .dynamic(false)
+        .tags("added[1.18.0]")
+        .buildOptional();
+
     public boolean isProfilingEnabled() {
         return profilingEnabled.get();
     }
@@ -205,5 +219,9 @@ public class ProfilingConfiguration extends ConfigurationOptionProvider {
 
     public boolean isBackupDiagnosticFiles() {
         return backupDiagnosticFiles.get();
+    }
+
+    public String getProfilerLibDirectory() {
+        return profilerLibDirectory.get().orElse(System.getProperty("java.io.tmpdir"));
     }
 }

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/SamplingProfiler.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/SamplingProfiler.java
@@ -259,7 +259,7 @@ public class SamplingProfiler extends AbstractLifecycleListener implements Runna
     public boolean onActivation(TraceContext activeSpan, @Nullable TraceContext previouslyActive) {
         if (profilingSessionOngoing) {
             if (previouslyActive == null) {
-                AsyncProfiler.getInstance(config).enableProfilingCurrentThread();
+                AsyncProfiler.getInstance(config.getProfilerLibDirectory()).enableProfilingCurrentThread();
             }
             boolean success = eventBuffer.tryPublishEvent(ACTIVATION_EVENT_TRANSLATOR, activeSpan, previouslyActive);
             if (!success && logger.isDebugEnabled()) {
@@ -284,7 +284,7 @@ public class SamplingProfiler extends AbstractLifecycleListener implements Runna
     public boolean onDeactivation(TraceContext activeSpan, @Nullable TraceContext previouslyActive) {
         if (profilingSessionOngoing) {
             if (previouslyActive == null) {
-                AsyncProfiler.getInstance(config).disableProfilingCurrentThread();
+                AsyncProfiler.getInstance(config.getProfilerLibDirectory()).disableProfilingCurrentThread();
             }
             boolean success = eventBuffer.tryPublishEvent(DEACTIVATION_EVENT_TRANSLATOR, activeSpan, previouslyActive);
             if (!success && logger.isDebugEnabled()) {
@@ -333,7 +333,7 @@ public class SamplingProfiler extends AbstractLifecycleListener implements Runna
     }
 
     private void profile(TimeDuration sampleRate, TimeDuration profilingDuration) throws Exception {
-        AsyncProfiler asyncProfiler = AsyncProfiler.getInstance(config);
+        AsyncProfiler asyncProfiler = AsyncProfiler.getInstance(config.getProfilerLibDirectory());
         try {
             String startCommand = "start,jfr,event=wall,cstack=n,interval=" + sampleRate.getMillis() + "ms,filter,file=" + jfrFile + ",safemode=" + config.getAsyncProfilerSafeMode();
             String startMessage = asyncProfiler.execute(startCommand);

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/SamplingProfiler.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/SamplingProfiler.java
@@ -259,7 +259,7 @@ public class SamplingProfiler extends AbstractLifecycleListener implements Runna
     public boolean onActivation(TraceContext activeSpan, @Nullable TraceContext previouslyActive) {
         if (profilingSessionOngoing) {
             if (previouslyActive == null) {
-                AsyncProfiler.getInstance().enableProfilingCurrentThread();
+                AsyncProfiler.getInstance(config).enableProfilingCurrentThread();
             }
             boolean success = eventBuffer.tryPublishEvent(ACTIVATION_EVENT_TRANSLATOR, activeSpan, previouslyActive);
             if (!success && logger.isDebugEnabled()) {
@@ -284,7 +284,7 @@ public class SamplingProfiler extends AbstractLifecycleListener implements Runna
     public boolean onDeactivation(TraceContext activeSpan, @Nullable TraceContext previouslyActive) {
         if (profilingSessionOngoing) {
             if (previouslyActive == null) {
-                AsyncProfiler.getInstance().disableProfilingCurrentThread();
+                AsyncProfiler.getInstance(config).disableProfilingCurrentThread();
             }
             boolean success = eventBuffer.tryPublishEvent(DEACTIVATION_EVENT_TRANSLATOR, activeSpan, previouslyActive);
             if (!success && logger.isDebugEnabled()) {
@@ -333,7 +333,7 @@ public class SamplingProfiler extends AbstractLifecycleListener implements Runna
     }
 
     private void profile(TimeDuration sampleRate, TimeDuration profilingDuration) throws Exception {
-        AsyncProfiler asyncProfiler = AsyncProfiler.getInstance();
+        AsyncProfiler asyncProfiler = AsyncProfiler.getInstance(config);
         try {
             String startCommand = "start,jfr,event=wall,cstack=n,interval=" + sampleRate.getMillis() + "ms,filter,file=" + jfrFile + ",safemode=" + config.getAsyncProfilerSafeMode();
             String startMessage = asyncProfiler.execute(startCommand);

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfiler.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfiler.java
@@ -39,7 +39,6 @@
  */
 package co.elastic.apm.agent.profiler.asyncprofiler;
 
-import co.elastic.apm.agent.profiler.ProfilingConfiguration;
 import co.elastic.apm.agent.util.IOUtils;
 
 import javax.annotation.Nullable;
@@ -49,7 +48,7 @@ import java.io.IOException;
 /**
  * Java API for in-process profiling. Serves as a wrapper around
  * async-profiler native library. This class is a singleton.
- * The first call to {@link #getInstance(ProfilingConfiguration config)} initiates loading of
+ * The first call to {@link #getInstance(String profilerLibDirectory)} initiates loading of
  * libasyncProfiler.so.
  * <p>
  * This is based on https://github.com/jvm-profiling-tools/async-profiler/blob/master/src/java/one/profiler/AsyncProfiler.java,
@@ -65,7 +64,7 @@ public class AsyncProfiler {
     private AsyncProfiler() {
     }
 
-    public static AsyncProfiler getInstance(ProfilingConfiguration config) {
+    public static AsyncProfiler getInstance(String profilerLibDirectory) {
         AsyncProfiler result = AsyncProfiler.instance;
         if (result != null) {
             return result;
@@ -73,16 +72,22 @@ public class AsyncProfiler {
         synchronized (AsyncProfiler.class) {
             if (instance == null) {
                 try {
-                    loadNativeLibrary(config.getProfilerLibDirectory());
+                    loadNativeLibrary(profilerLibDirectory);
                 } catch (UnsatisfiedLinkError e) {
                     throw new IllegalStateException(String.format("It is likely that %s is not an executable location. Consider setting " +
                         "the profiling_inferred_spans_lib_directory property to a directory on a partition that allows execution",
-                        config.getProfilerLibDirectory()), e);
+                        profilerLibDirectory), e);
                 }
 
                 instance = new AsyncProfiler();
             }
             return instance;
+        }
+    }
+
+    static void reset() {
+        synchronized (AsyncProfiler.class) {
+            instance = null;
         }
     }
 

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfiler.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfiler.java
@@ -39,6 +39,7 @@
  */
 package co.elastic.apm.agent.profiler.asyncprofiler;
 
+import co.elastic.apm.agent.profiler.ProfilingConfiguration;
 import co.elastic.apm.agent.util.IOUtils;
 
 import javax.annotation.Nullable;
@@ -48,7 +49,7 @@ import java.io.IOException;
 /**
  * Java API for in-process profiling. Serves as a wrapper around
  * async-profiler native library. This class is a singleton.
- * The first call to {@link #getInstance()} initiates loading of
+ * The first call to {@link #getInstance(ProfilingConfiguration config)} initiates loading of
  * libasyncProfiler.so.
  * <p>
  * This is based on https://github.com/jvm-profiling-tools/async-profiler/blob/master/src/java/one/profiler/AsyncProfiler.java,
@@ -64,14 +65,20 @@ public class AsyncProfiler {
     private AsyncProfiler() {
     }
 
-    public static AsyncProfiler getInstance() {
+    public static AsyncProfiler getInstance(ProfilingConfiguration config) {
         AsyncProfiler result = AsyncProfiler.instance;
         if (result != null) {
             return result;
         }
         synchronized (AsyncProfiler.class) {
             if (instance == null) {
-                loadNativeLibrary();
+                try {
+                    loadNativeLibrary(config.getProfilerLibDirectory());
+                } catch (UnsatisfiedLinkError e) {
+                    throw new IllegalStateException(String.format("Is is likely that %s is not an executable location. Consider setting " +
+                        "the profiling_inferred_spans_lib_directory property to a directory on a partition that allows execution",
+                        config.getProfilerLibDirectory()), e);
+                }
 
                 instance = new AsyncProfiler();
             }
@@ -79,9 +86,10 @@ public class AsyncProfiler {
         }
     }
 
-    private static void loadNativeLibrary() {
+    private static void loadNativeLibrary(String libraryDirectory) {
         String libraryName = getLibraryFileName();
-        File file = IOUtils.exportResourceToTemp("asyncprofiler/" + libraryName + ".so", libraryName, ".so");
+        File file = IOUtils.exportResourceToDirectory("asyncprofiler/" + libraryName + ".so", libraryDirectory,
+            libraryName, ".so");
         System.load(file.getAbsolutePath());
     }
 

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfiler.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfiler.java
@@ -75,7 +75,7 @@ public class AsyncProfiler {
                 try {
                     loadNativeLibrary(config.getProfilerLibDirectory());
                 } catch (UnsatisfiedLinkError e) {
-                    throw new IllegalStateException(String.format("Is is likely that %s is not an executable location. Consider setting " +
+                    throw new IllegalStateException(String.format("It is likely that %s is not an executable location. Consider setting " +
                         "the profiling_inferred_spans_lib_directory property to a directory on a partition that allows execution",
                         config.getProfilerLibDirectory()), e);
                 }

--- a/apm-agent-plugins/apm-profiling-plugin/src/test/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfilerTest.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/test/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfilerTest.java
@@ -24,55 +24,37 @@
  */
 package co.elastic.apm.agent.profiler.asyncprofiler;
 
-import co.elastic.apm.agent.configuration.SpyConfiguration;
-import co.elastic.apm.agent.profiler.ProfilingConfiguration;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
-import org.stagemonitor.configuration.ConfigurationRegistry;
+import org.junit.jupiter.api.io.TempDir;
 
 import java.io.File;
 import java.io.FilenameFilter;
-import java.io.IOException;
-import java.lang.reflect.Field;
-import java.nio.file.Files;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.mockito.Mockito.when;
 
 public class AsyncProfilerTest {
 
-    private ProfilingConfiguration profilerConfig;
-
     @BeforeEach
-    void setUp() throws ReflectiveOperationException {
-        ConfigurationRegistry config = SpyConfiguration.createSpyConfig();
-        profilerConfig = config.getConfig(ProfilingConfiguration.class);
-
-        // Ensure that the singleton AsyncProfiler is reset so a new instance is created for each test
-        Field instance = AsyncProfiler.class.getDeclaredField("instance");
-        instance.setAccessible(true);
-        instance.set(null, null);
+    void setUp() {
+        AsyncProfiler.reset();
     }
 
     @Test
     void testShouldCopyLibToTempDirectory() {
-        AsyncProfiler.getInstance(profilerConfig);
+        String defaultTempDirectory = System.getProperty("java.io.tmpdir");
+        AsyncProfiler.getInstance(defaultTempDirectory);
 
-        File libDirectory = new File(profilerConfig.getProfilerLibDirectory());
+        File libDirectory = new File(defaultTempDirectory);
         File[] libasyncProfilers = libDirectory.listFiles(getLibasyncProfilerFilenameFilter());
         assertThat(libasyncProfilers).hasSizeGreaterThanOrEqualTo(1);
     }
 
     @Test
-    void testShouldCopyLibToSpecifiedDirectory() throws IOException {
-        File parentDirectory = Files.createTempDirectory(null).toFile();
-        parentDirectory.deleteOnExit();
-        when(profilerConfig.getProfilerLibDirectory()).thenReturn(parentDirectory.getAbsolutePath());
-        System.out.println(parentDirectory.getAbsolutePath());
+    void testShouldCopyLibToSpecifiedDirectory(@TempDir File nonDefaultTempDirectory) {
+        AsyncProfiler.getInstance(nonDefaultTempDirectory.getAbsolutePath());
 
-        AsyncProfiler.getInstance(profilerConfig);
-
-        File[] libasyncProfilers = parentDirectory.listFiles(getLibasyncProfilerFilenameFilter());
+        File[] libasyncProfilers = nonDefaultTempDirectory.listFiles(getLibasyncProfilerFilenameFilter());
         assertThat(libasyncProfilers).hasSizeGreaterThanOrEqualTo(1);
     }
 

--- a/apm-agent-plugins/apm-profiling-plugin/src/test/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfilerTest.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/test/java/co/elastic/apm/agent/profiler/asyncprofiler/AsyncProfilerTest.java
@@ -1,0 +1,82 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.profiler.asyncprofiler;
+
+import co.elastic.apm.agent.configuration.SpyConfiguration;
+import co.elastic.apm.agent.profiler.ProfilingConfiguration;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.stagemonitor.configuration.ConfigurationRegistry;
+
+import java.io.File;
+import java.io.FilenameFilter;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.nio.file.Files;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+public class AsyncProfilerTest {
+
+    private ProfilingConfiguration profilerConfig;
+
+    @BeforeEach
+    void setUp() throws ReflectiveOperationException {
+        ConfigurationRegistry config = SpyConfiguration.createSpyConfig();
+        profilerConfig = config.getConfig(ProfilingConfiguration.class);
+
+        // Ensure that the singleton AsyncProfiler is reset so a new instance is created for each test
+        Field instance = AsyncProfiler.class.getDeclaredField("instance");
+        instance.setAccessible(true);
+        instance.set(null, null);
+    }
+
+    @Test
+    void testShouldCopyLibToTempDirectory() {
+        AsyncProfiler.getInstance(profilerConfig);
+
+        File libDirectory = new File(profilerConfig.getProfilerLibDirectory());
+        File[] libasyncProfilers = libDirectory.listFiles(getLibasyncProfilerFilenameFilter());
+        assertThat(libasyncProfilers).hasSizeGreaterThanOrEqualTo(1);
+    }
+
+    @Test
+    void testShouldCopyLibToSpecifiedDirectory() throws IOException {
+        File parentDirectory = Files.createTempDirectory(null).toFile();
+        parentDirectory.deleteOnExit();
+        when(profilerConfig.getProfilerLibDirectory()).thenReturn(parentDirectory.getAbsolutePath());
+        System.out.println(parentDirectory.getAbsolutePath());
+
+        AsyncProfiler.getInstance(profilerConfig);
+
+        File[] libasyncProfilers = parentDirectory.listFiles(getLibasyncProfilerFilenameFilter());
+        assertThat(libasyncProfilers).hasSizeGreaterThanOrEqualTo(1);
+    }
+
+    private FilenameFilter getLibasyncProfilerFilenameFilter() {
+        return (dir, name) -> name.startsWith("libasyncProfiler") && name.endsWith(".so");
+    }
+}

--- a/docs/configuration.asciidoc
+++ b/docs/configuration.asciidoc
@@ -1874,7 +1874,7 @@ If unset, the value of the "java.io.tmpdir" System property will be used.
 [options="header"]
 |============
 | Default                          | Type                | Dynamic
-| `<none>` | Optional | false
+| `<none>` | String | false
 |============
 
 
@@ -3222,7 +3222,7 @@ The default unit for this option is `ms`.
 # If unset, the value of the "java.io.tmpdir" System property will be used.
 #
 # This setting can not be changed at runtime. Changes require a restart of the application.
-# Type: Optional
+# Type: String
 # Default value: 
 #
 # profiling_inferred_spans_lib_directory=

--- a/docs/configuration.asciidoc
+++ b/docs/configuration.asciidoc
@@ -1864,9 +1864,9 @@ Prepending an element with `(?-i)` makes the matching case sensitive.
 ==== `profiling_inferred_spans_lib_directory` (added[1.18.0])
 
 Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library is exported to a temporary location and loaded by the JVM.
-The partition backing this location must be executable, however in some server-hardened environments, noexec may be set on the standard /tmp partition, leading to java.lang.UnsatisfiedLinkError errors.
-Set this property to an alternative directory (e.g. '/var/tmp') to resolve this.
-If unset, the value of the "java.io.tmpdir" System property will be used.
+The partition backing this location must be executable, however in some server-hardened environments, noexec may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.
+Set this property to an alternative directory (e.g. `/var/tmp`) to resolve this.
+If unset, the value of the `java.io.tmpdir` System property will be used.
 
 
 
@@ -3217,9 +3217,9 @@ The default unit for this option is `ms`.
 # profiling_inferred_spans_excluded_classes=(?-i)java.*,(?-i)javax.*,(?-i)sun.*,(?-i)com.sun.*,(?-i)jdk.*,(?-i)org.apache.tomcat.*,(?-i)org.apache.catalina.*,(?-i)org.apache.coyote.*,(?-i)org.jboss.as.*,(?-i)org.glassfish.*,(?-i)org.eclipse.jetty.*,(?-i)com.ibm.websphere.*,(?-i)io.undertow.*
 
 # Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library is exported to a temporary location and loaded by the JVM.
-# The partition backing this location must be executable, however in some server-hardened environments, noexec may be set on the standard /tmp partition, leading to java.lang.UnsatisfiedLinkError errors.
-# Set this property to an alternative directory (e.g. '/var/tmp') to resolve this.
-# If unset, the value of the "java.io.tmpdir" System property will be used.
+# The partition backing this location must be executable, however in some server-hardened environments, noexec may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.
+# Set this property to an alternative directory (e.g. `/var/tmp`) to resolve this.
+# If unset, the value of the `java.io.tmpdir` System property will be used.
 #
 # This setting can not be changed at runtime. Changes require a restart of the application.
 # Type: String

--- a/docs/configuration.asciidoc
+++ b/docs/configuration.asciidoc
@@ -149,6 +149,7 @@ Click on a key to get more information.
 ** <<config-profiling-inferred-spans-min-duration>>
 ** <<config-profiling-inferred-spans-included-classes>>
 ** <<config-profiling-inferred-spans-excluded-classes>>
+** <<config-profiling-inferred-spans-lib-directory>>
 * <<config-reporter>>
 ** <<config-secret-token>>
 ** <<config-api-key>>
@@ -1857,6 +1858,32 @@ Prepending an element with `(?-i)` makes the matching case sensitive.
 | `elastic.apm.profiling_inferred_spans_excluded_classes` | `profiling_inferred_spans_excluded_classes` | `ELASTIC_APM_PROFILING_INFERRED_SPANS_EXCLUDED_CLASSES`
 |============
 
+// This file is auto generated. Please make your changes in *Configuration.java (for example CoreConfiguration.java) and execute ConfigurationExporter
+[float]
+[[config-profiling-inferred-spans-lib-directory]]
+==== `profiling_inferred_spans_lib_directory` (added[1.18.0])
+
+Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library is exported to a temporary location and loaded by the JVM.
+The partition backing this location must be executable, however in some server-hardened environments, noexec may be set on the standard /tmp partition, leading to java.lang.UnsatisfiedLinkError errors.
+Set this property to an alternative directory (e.g. '/var/tmp') to resolve this.
+If unset, the value of the "java.io.tmpdir" System property will be used.
+
+
+
+
+[options="header"]
+|============
+| Default                          | Type                | Dynamic
+| `<none>` | Optional | false
+|============
+
+
+[options="header"]
+|============
+| Java System Properties      | Property file   | Environment
+| `elastic.apm.profiling_inferred_spans_lib_directory` | `profiling_inferred_spans_lib_directory` | `ELASTIC_APM_PROFILING_INFERRED_SPANS_LIB_DIRECTORY`
+|============
+
 [[config-reporter]]
 === Reporter configuration options
 // This file is auto generated. Please make your changes in *Configuration.java (for example CoreConfiguration.java) and execute ConfigurationExporter
@@ -3188,6 +3215,17 @@ The default unit for this option is `ms`.
 # Default value: (?-i)java.*,(?-i)javax.*,(?-i)sun.*,(?-i)com.sun.*,(?-i)jdk.*,(?-i)org.apache.tomcat.*,(?-i)org.apache.catalina.*,(?-i)org.apache.coyote.*,(?-i)org.jboss.as.*,(?-i)org.glassfish.*,(?-i)org.eclipse.jetty.*,(?-i)com.ibm.websphere.*,(?-i)io.undertow.*
 #
 # profiling_inferred_spans_excluded_classes=(?-i)java.*,(?-i)javax.*,(?-i)sun.*,(?-i)com.sun.*,(?-i)jdk.*,(?-i)org.apache.tomcat.*,(?-i)org.apache.catalina.*,(?-i)org.apache.coyote.*,(?-i)org.jboss.as.*,(?-i)org.glassfish.*,(?-i)org.eclipse.jetty.*,(?-i)com.ibm.websphere.*,(?-i)io.undertow.*
+
+# Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library is exported to a temporary location and loaded by the JVM.
+# The partition backing this location must be executable, however in some server-hardened environments, noexec may be set on the standard /tmp partition, leading to java.lang.UnsatisfiedLinkError errors.
+# Set this property to an alternative directory (e.g. '/var/tmp') to resolve this.
+# If unset, the value of the "java.io.tmpdir" System property will be used.
+#
+# This setting can not be changed at runtime. Changes require a restart of the application.
+# Type: Optional
+# Default value: 
+#
+# profiling_inferred_spans_lib_directory=
 
 ############################################
 # Reporter                                 #

--- a/docs/configuration.asciidoc
+++ b/docs/configuration.asciidoc
@@ -1866,7 +1866,7 @@ Prepending an element with `(?-i)` makes the matching case sensitive.
 Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library is exported to a temporary location and loaded by the JVM.
 The partition backing this location must be executable, however in some server-hardened environments, noexec may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.
 Set this property to an alternative directory (e.g. `/var/tmp`) to resolve this.
-If unset, the value of the `java.io.tmpdir` System property will be used.
+If unset, the value of the `java.io.tmpdir` system property will be used.
 
 
 
@@ -3219,7 +3219,7 @@ The default unit for this option is `ms`.
 # Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library is exported to a temporary location and loaded by the JVM.
 # The partition backing this location must be executable, however in some server-hardened environments, noexec may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.
 # Set this property to an alternative directory (e.g. `/var/tmp`) to resolve this.
-# If unset, the value of the `java.io.tmpdir` System property will be used.
+# If unset, the value of the `java.io.tmpdir` system property will be used.
 #
 # This setting can not be changed at runtime. Changes require a restart of the application.
 # Type: String

--- a/docs/configuration.asciidoc
+++ b/docs/configuration.asciidoc
@@ -1864,7 +1864,7 @@ Prepending an element with `(?-i)` makes the matching case sensitive.
 ==== `profiling_inferred_spans_lib_directory` (added[1.18.0])
 
 Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library is exported to a temporary location and loaded by the JVM.
-The partition backing this location must be executable, however in some server-hardened environments, noexec may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.
+The partition backing this location must be executable, however in some server-hardened environments, `noexec` may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.
 Set this property to an alternative directory (e.g. `/var/tmp`) to resolve this.
 If unset, the value of the `java.io.tmpdir` system property will be used.
 
@@ -3217,7 +3217,7 @@ The default unit for this option is `ms`.
 # profiling_inferred_spans_excluded_classes=(?-i)java.*,(?-i)javax.*,(?-i)sun.*,(?-i)com.sun.*,(?-i)jdk.*,(?-i)org.apache.tomcat.*,(?-i)org.apache.catalina.*,(?-i)org.apache.coyote.*,(?-i)org.jboss.as.*,(?-i)org.glassfish.*,(?-i)org.eclipse.jetty.*,(?-i)com.ibm.websphere.*,(?-i)io.undertow.*
 
 # Profiling requires that the https://github.com/jvm-profiling-tools/async-profiler[async-profiler] shared library is exported to a temporary location and loaded by the JVM.
-# The partition backing this location must be executable, however in some server-hardened environments, noexec may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.
+# The partition backing this location must be executable, however in some server-hardened environments, `noexec` may be set on the standard `/tmp` partition, leading to `java.lang.UnsatisfiedLinkError` errors.
 # Set this property to an alternative directory (e.g. `/var/tmp`) to resolve this.
 # If unset, the value of the `java.io.tmpdir` system property will be used.
 #

